### PR TITLE
Add reworked tests for Maternity Paternity Pay Leave

### PR DESCRIPTION
**Pull request by gclssvglx**
[Trello](https://trello.com/c/3LmQ9I5i/628-smart-answer-flow-tests-maternity-paternity-pay-leave)

This converts the Plan Adoption Leave tests to use the node based testing
approach introduced in [ADR-3][1].

[1]: https://github.com/alphagov/smart-answers/blob/main/docs/arch/003-testing-flows.md

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

### CODE REVIEW

File: test/flows/maternity_paternity_pay_leave_flow_test.rb [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+                    due_date: "2016-1-1",
+                    employment_status_of_mother: "employee"
+    end
+
+    should "render the question" do
+      assert_rendered_question
+    end
+
+    context "next_node" do
+      should "have a next node of mother_started_working_before_continuity_start_date for any response" do
```
Comment (leenagupte) [inline test/flows/maternity_paternity_pay_leave_flow_test.rb:91]: So here you probably do want a test for each employment status of the mother, as it does affect the flow.
```suggestion
      should "have a next node of mother_started_working_before_continuity_start_date for any response when the mother is an employee" do
```

File: test/flows/maternity_paternity_pay_leave_flow_test.rb [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+                    mother_started_working_before_continuity_start_date: "yes",
+                    mother_still_working_on_continuity_end_date: "yes"
+    end
+
+    should "render the question" do
+      assert_rendered_question
+    end
+
+    context "next_node" do
+      should "have a next node of partner_started_working_before_continuity_start_date for a 'yes' response" do
```
Comment (leenagupte) [inline test/flows/maternity_paternity_pay_leave_flow_test.rb:159]: I think this needs something in the description to say what makes it different from the other "yes" response test.
```suggestion
      should "have a next node of partner_started_working_before_continuity_start_date for a 'yes' response if partner is an employee" do
```

**Review by leenagupte — COMMENTED** [review]
Graham, the structure of the tests is looking good. I think they're just missing a little extra description in the test name to make them a bit easier to follow. I've added a couple of examples in-line. Otherwise I think this is on the right track. 

Well done for getting your head around this! ⭐

**Comment by gclssvglx** [comment]
Great suggestions thanks @leenagupte 👍  I'll update and fix the existing tests.

**Comment by leenagupte** [comment]
After another marathon reviewing session, every up to `partner_earned_more_than_lower_earnings_limit` has been approved.

Things left to test:
question: `partner_earned_more_than_lower_earnings_limit`
outcome tests.

**Comment by leenagupte** [comment]
Current progress: Reviewing tests for line [369](https://github.com/alphagov/smart-answers/blob/main/app/flows/maternity_paternity_pay_leave_flow.rb#L369)

**Comment by leenagupte** [comment]
Current progress: https://github.com/alphagov/smart-answers/blob/main/app/flows/maternity_paternity_pay_leave_flow.rb#L490

**Comment by leenagupte** [comment]
The tests look good to me! Waiting to commits to be fixed up and commented out tests to be removed, and then we're good to go 🎉

**Comment by gclssvglx** [comment]
> The tests look good to me! Waiting to commits to be fixed up and commented out tests to be removed, and then we're good to go 🎉

All comments removed, tests added for 2022 outcomes, fixed up.

**Review by leenagupte — APPROVED** [review]
Well done! 👏 🏅 ⭐ 💯 👍 🎉 🏆 
The tests look great. The flow you had to work with is absolutely terrible, I've never seen logic so convoluted, but you stuck with it. Your hard work means that it will be possible to refactor this flow in the future and have confidence that the logic works. Thank you!

**Pull request merged**